### PR TITLE
[storage] more pruner fixes

### PR DESCRIPTION
**Pull request by msmouse**
### Test Plan

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/3327)

### CODE REVIEW

File: storage/aptosdb/src/pruner/state_pruner_worker.rs [hunk 1]
```diff
@@ -62,8 +62,9 @@ impl StatePrunerWorker {
     }
 
     pub fn set_target_db_version(&self, target_db_version: Version) {
-        assert!(target_db_version >= self.pruner.target_version());
```
Comment (zcchahaha) [inline storage/aptosdb/src/pruner/state_pruner_worker.rs:65]: Why do we remove this assertion?

File: storage/aptosdb/src/pruner/state_store/mod.rs [hunk 2]
```diff
@@ -206,36 +206,23 @@ impl StateMerklePruner {
...
             }
+            break;
         }
 
-        // This is to deal with the case where we reaches the end of the indices regardless of
-        // whether we have `num_items` in `indices`.
-        let mut is_end_of_target_version = true;
-        if let Some(last_index) = indices.last() {
-            is_end_of_target_version = last_index.stale_since_version == target_version;
-        }
```
Comment (zcchahaha) [inline storage/aptosdb/src/pruner/state_store/mod.rs:238]: what is the bug of the previous implementation?

File: storage/aptosdb/src/lib.rs [hunk 3]
```diff
@@ -1558,7 +1554,8 @@ impl DbWriter for AptosDB {
                 let last_version = first_version + num_txns - 1;
                 COMMITTED_TXNS.inc_by(num_txns);
                 LATEST_TXN_VERSION.set(last_version as i64);
-                self.set_pruner_target_version(last_version);
+                self.ledger_pruner
```
Comment (zcchahaha) [inline storage/aptosdb/src/lib.rs:1557]: Should we add some comments on when the states merkle tree will be pruned?

File: storage/aptosdb/src/pruner/state_store/mod.rs (hunk 2 above)
Comment (msmouse) [reply storage/aptosdb/src/pruner/state_store/mod.rs:238]: The while loop on line 210 exits when batch_size is met, but there can potentially still be indices for the target version in the CF that didn't fit the batch. In that case line 226 is a true but line 228 is a false, so we are assuming `is_end_of_target_version = true` at line #237 and return it on line 239. which is wrong.

File: storage/aptosdb/src/pruner/state_pruner_worker.rs (hunk 1 above)
Comment (msmouse) [reply storage/aptosdb/src/pruner/state_pruner_worker.rs:65]: The test I added in storage/aptosdb/src/aptosdb_test.rs, line 237 I "skip ahead" the pruner a bit to get a "clean DB" with all nodes referenced by some root, in order to test all that need to be pruned are marked by the stale node index. However that'll make this assertion fail.

**Pull request merged**